File: SRTF algorithm/srtf.py
#
# Created by arogya.Gurkha on 2020-04-19.
#
import operator


class Process:
    def __init__(self, process_number, arrival_time, use_time,
                 waiting_time):  # will contain the process data from the input file
        self.process_number = process_number
        self.arrival_time = arrival_time
        self.use_time = use_time
        self.waiting_time = waiting_time  # later when the process might need to wait


def calculation(container, count):
    process_id, r_process, sum_wait_time = -1, count, 0  # current process id, remaining processes, total waiting time
    i = 0
    while r_process > 0:  # if there are processes left to be processed
        container.sort(key=operator.attrgetter('arrival_time', 'use_time',
                                               'process_number'))  # sort with arrival time, use time and process num
        min_time = 2100000  # min time for processing

        for j in range(0, count):
            if container[j].arrival_time > i:  # if the process arrives too late
                break
            if container[j].use_time > 0 and container[j].arrival_time <= i:
                # if the process still needs time and has already arrived
                if min_time > container[j].use_time:
                    min_time = container[j].use_time  # change the small amount of time
                    process_id = j  # current iteration will be the process to have its use time reduced

        if process_id != -1:  # if the process is valid
            container[process_id].use_time -= 1  # then reduce the use time
            if container[process_id].use_time == 0:  # if the process is finished then reduce the number of process
                r_process -= 1

        for j in range(0, count):
            if j != process_id and container[j].arrival_time <= i and container[j].use_time > 0:
                # if the current iteration isn't the current process, if the process has already arrived and
                # still needs to be processed
                container[j].waiting_time += 1  # increase the waiting time

        i += 1  # increase the time by 1

    for i in range(0, count):
        sum_wait_time += container[i].waiting_time  # add all the waiting times

    return sum_wait_time


if __name__ == '__main__':
    process_container = []
    with open('srtf.inp') as inp, open('srtf.out', 'w') as out:

        process_count = int(inp.readline())

        for line in inp:
            num, arr, dur = map(int, line.split())
            process_container.append(Process(num, arr, dur, 0))

        out.write(str(calculation(process_container, process_count)))  # output

File: SRTF algorithm/srtf.cpp
//
// Created by arogya.Gurkha on 2020-04-19.
//

#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

#define _min 21000000

std::ifstream inp("srtf.inp");
std::ofstream out("srtf.out");

typedef struct Process { // will contain the process data from the input file
    int processNumber = 0;
    int arrivalTime = 0;
    int useTime = 0;
    int waitTime = 0; // later when the process might need to wait
};

bool sortingArrival(const Process &p1, const Process &p2) {

    if (p1.arrivalTime < p2.arrivalTime) return true; // first sort by arrival
    if (p1.arrivalTime > p2.arrivalTime) return false;

    if (p1.useTime < p2.useTime) return true; // second by use time
    if (p1.useTime > p2.useTime) return false;

    return p1.processNumber < p2.processNumber; // and lastly with process number
}

void calculation(std::vector<Process> &vec, int count) {

    int processID = -1, r_process = count; // current processID and remaining process

    for (int i = 0; r_process > 0; i++) {
        std::sort(vec.begin(), vec.end(), sortingArrival);
        int min = _min; // smallest use time

        for (int j = 0; j < count; j++) {
            Process *c_process = &vec[j]; // current process
            if (c_process->arrivalTime > i) break; // if the current process arrives later than current time then break
            if (c_process->useTime > 0 && c_process->arrivalTime <= i) { // if there is time left in the process and arrival time is in present
                if (min > c_process->useTime) { // check if the current process' use time is the min
                    min = c_process->useTime; // if yes then the min is current process
                    processID = j; // save the current process id in processID
                }
            }
        }
        if (processID != -1) { // if the processID is valid
            vec[processID].useTime--; // then reduce the use time by 1
            if (vec[processID].useTime == 0) // check if use time is 0 for the process
                r_process--; // if yes then that process can be removed
        }
        for (int j = 0; j < count; j++) {
            if (j != processID && vec[j].arrivalTime <= i && vec[j].useTime > 0) // j is a process that is selected if it fits all the criteria
                vec[j].waitTime++; // then increase the waiting time for j
        }
    }
}

int main() {

    std::vector<Process> processContainer;

    int processCount;
    inp >> processCount;

    int totalTime = 0, sum_waitTime = 0; // total time taken by all the processes

    for (int i = 0; i < processCount; i++) {
        Process process;
        inp >> process.processNumber >> process.arrivalTime >> process.useTime;

        if (totalTime < process.arrivalTime)
            totalTime += process.arrivalTime - totalTime;

        totalTime += process.useTime;

        processContainer.push_back(process); // get data from the input file and store them all in a vector
    }

    calculation(processContainer, processCount);

    for (int i = 0; i < processCount; i++)
        sum_waitTime += processContainer[i].waitTime; // add all the wait times to get the final answer

    out << sum_waitTime;
}
